currency_rate_check/models/account_move_line_currency_check.py

```diff
@@ -107,41 +115,41 @@ def action_refresh_view(self, date_from, date_to, line_limit, ref):
                         line.credit, abs(amount), precision_digits=precision
                     )
                 ):
-                    lines_to_check |= line
-        if lines_to_check:
-            drop_view_if_exists(self._cr, 'account_move_line_currency_check')
-            self.env.cr.execute(
-                """
-                CREATE OR REPLACE VIEW account_move_line_currency_check AS (
-                    SELECT aml.id AS id,
-                    aml.id AS line_id,
-                    rate_partner.id AS rate_partner,
-                    rate_company.id AS rate_company
-                    FROM account_move_line AS aml
-                    INNER JOIN res_currency_rate AS rate_partner
-                    ON rate_partner.name = aml.date
-                    AND rate_partner.company_id = %(company_id)s
-                    AND rate_partner.currency_id = aml.currency_id
-                    INNER JOIN res_currency_rate AS rate_company
-                    ON rate_company.name = aml.date
-                    AND rate_company.company_id = %(company_id)s
-                    AND rate_company.currency_id = aml.company_currency_id
-                    WHERE aml.company_id = %(company_id)s
-                    AND aml.id IN %(lines_to_check)s
-                )
-                """, {
-                    'company_id': self.env.user.company_id.id,
-                    'lines_to_check': tuple(lines_to_check.ids),
-                }
-            )
-        return lines_to_check
+                    domain = [
+                        ('name', '=', line.date),
+                        ('company_id', '=', self.env.user.company_id.id),
+                    ]
+                    rate_partner = self.env['res.currency.rate'].search(
+                        domain + [('currency_id', '=', line.currency_id.id)]
+                    )
+                    rate_company = self.env['res.currency.rate'].search(
+                        domain + [('currency_id', '=', line.company_currency_id.id)]
+                    )
+                    values = {
+                        'line_id': line.id,
+                        'rate_partner': rate_partner.id,
+                        'rate_company': rate_company.id,
+                    }
+                    check_line = self.search([('line_id', '=', line.id)])
+                    if len(check_line) > 0:
+                        check_line.write(values)
+                        write_values.append(values)
+                    else:
+                        create_values.append(values)
+        if create_values:
+            self.create(create_values)
```

Comment:
isn't this horribly slow? Why don't you actually do an [upsert with SQL](https://www.postgresql.org/docs/9.5/sql-insert.html#SQL-ON-CONFLICT)?

Reply:
I don't think it has to be fast as this check only has to be done if something went wrong. Locally it runs fast enough (a few seconds) that I wasn't worried about performance. Also: https://github.com/VanMoof/currency/pull/2#discussion_r740834776

Reply:
fine with me